File: java/bank-account/src/main/java/BankAccount.java
class BankAccount {

    private int balance;
    private boolean isOpen;

    public BankAccount() {
        this.balance = 0;
        this.isOpen = false;
    }

    void open() throws BankAccountActionInvalidException {
        if (this.isOpen) {
            throw new BankAccountActionInvalidException("Account already open");
        }

        this.isOpen = true;
    }

    void close() throws BankAccountActionInvalidException {
        if (!this.isOpen) {
            throw new BankAccountActionInvalidException("Account not open");
        }

        this.balance = 0;
        this.isOpen = false;
    }

    synchronized int getBalance() throws BankAccountActionInvalidException {
        if (!this.isOpen) {
            throw new BankAccountActionInvalidException("Account closed");
        }

        return this.balance;
    }

    synchronized void deposit(int amount) throws BankAccountActionInvalidException {
        if (!this.isOpen) {
            throw new BankAccountActionInvalidException("Account closed");
        }

        if (amount < 0) {
            throw new BankAccountActionInvalidException("Cannot deposit or withdraw negative amount");
        }

        this.balance += amount;
    }

    synchronized void withdraw(int amount) throws BankAccountActionInvalidException {
        if (!this.isOpen) {
            throw new BankAccountActionInvalidException("Account closed");
        }

        if (amount > this.balance) {
            throw new BankAccountActionInvalidException("Cannot withdraw more money than is currently in the account");
        }

        if (amount < 0) {
            throw new BankAccountActionInvalidException("Cannot deposit or withdraw negative amount");
        }

        this.balance -= amount;
    }

}

File: python/bank-account/bank_account.py
class BankAccount:
    def __init__(self):
        self.balance = 0
        self.account_open = False

    def get_balance(self):
        if not self.account_open:
            raise ValueError("account not open")

        return self.balance

    def open(self):
        if self.account_open:
            raise ValueError("account already open")

        self.account_open = True

    def deposit(self, amount):
        if not self.account_open:
            raise ValueError("account not open")

        if amount <= 0:
            raise ValueError("amount must be greater than 0")

        self.balance += amount

    def withdraw(self, amount):
        if not self.account_open:
            raise ValueError("account not open")

        if amount > self.balance:
            raise ValueError("amount must be less than balance")

        if amount <= 0:
            raise ValueError("amount must be greater than 0")

        self.balance -= amount

    def close(self):
        if not self.account_open:
            raise ValueError("account not open")

        self.balance = 0
        self.account_open = False
